File: algorithm/sort/选择排序.py
import random


def get_random_list(length: int=10, start: int=0, end: int=100) -> list:
    """获取随机数列表"""
    result = []
    for _ in range(length):
        result.append(random.choice(range(start, end)))
    return result


def my_sort(origin_list: list) -> list:
    """排序算法"""
    for i in range(len(origin_list)-1):
        min = i 
        for j in range(i, len(origin_list)):
            if origin_list[min] > origin_list[j]:
                min = j
        if min != i:
            origin_list[i], origin_list[min] = origin_list[min], origin_list[i]
    return origin_list


if __name__ == "__main__":
    for i in range(20):
        length = random.choice(range(0, 20))
        tmp = get_random_list(length)
        if my_sort(tmp) != sorted(tmp):
            print('排序失败', tmp)
            break

File: algorithm/sort/选择排序.go
package main

import (
	"fmt"
	"math/rand"
	"sort"
	"time"
)

// get_rand_slice 获取指定长度和指定范围的随机切片
func get_rand_slice(length, start, end int) (result []int) {
	for i := 0; i < length; i++ {
		item := rand.Intn(end-start) + start
		result = append(result, item)
	}
	return
}


// mySort 选择排序golang实现
func mySort(origin []int) []int {
	length := len(origin)
	for i:=0; i<length; i++{
		min := i
		for j:=i; j<length; j++{
			if origin[j] < origin[min]{
				min = j
			}
		}
		if min != i {
			origin[i], origin[min] = origin[min], origin[i]
		}
	}
	return origin
}


// compare 排序结果比较器
func compare(slice1, slice2 []int) bool {
	len_slice1 := len(slice1)
	len_slice2 := len(slice2)
	if len_slice1 != len_slice2{
		return false
	}
	length := len_slice1
	for i:=0; i<length; i++{
		if slice1[i] != slice2[i]{
			return false
		}
	}
	return true
}


func main(){
	for i:=0; i<20; i++{
		// 设置随机数种子，保证每次产生的数都不一样
		rand.Seed(time.Now().UnixNano())
		length := rand.Intn(20)
		origin := get_rand_slice(length, 0, 100)
		fmt.Printf("%v", origin)
		my_sort := mySort(origin[:])
		build_in_sort := origin[:]
		sort.Ints(build_in_sort)
		if result := compare(my_sort, build_in_sort); result==false{
			fmt.Print(result)
			fmt.Print("排序失败")
			break;
		}
	}
	fmt.Println("排序成功")
}
